### PR TITLE
KOA-4826 Generate compose typography styles

### PR DESCRIPTION
**Pull request by marianeum**
Generate the compose typography styles, rather than creating them programmatically.

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] `UNRELEASED.md`
+ [x] Tests
+ [x] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+  fun String.toSemanticName() =
+    CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, this.replace('-', '_'))
+
+  fun BpkTextStyles.toTextStyleClass(): TypeSpec {
+
+    fun BpkTextStyles.toConstructorFunction(): FunSpec {
+
+      fun BpkTextStyleModel.toParameter(): String =
+        "%textStyle:T(\n" +
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt:123]: unfortunately there's no nice helper methods for the constructor function itself, so we have to build this all ourselves. the `%x:T` syntax is for named placeholders, so we can refer to class names etc

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+          .defaultValue("%T.SansSerif", FontFamilyClass)
+          .build())
+        .callThisConstructor(CodeBlock.builder().addNamed(joinToString(
+          prefix = "\n  ",
+          separator = ",\n  ",
+          postfix = ",\n",
+        ) {
+          "${it.name.toSemanticName()} = ${it.toParameter()}"
+        }, mapOf(
+          "textStyle" to TextStyleClass,
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt:146]: that's the mapping from `x` to the class for imports etc

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+import com.google.common.base.CaseFormat
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.ParameterSpec
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeSpec
+
+interface BpkTextStyles : List<BpkTextStyleModel>
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt:29]: most of this is very closely aligned with `BpkColors` (at least the semantic version) as the requirements are quite similar

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/tokens/BpkTypography.kt [hunk 4]
```diff
@@ -1,82 +1,65 @@
...
-      lineHeight = BpkLineHeight.`8xl`,
-      letterSpacing = BpkLetterSpacing.Tight,
-      fontFamily = defaultFontFamily,
-    ),
-    hero2 = TextStyle(
-      fontWeight = FontWeight.Bold,
-      fontSize = BpkFontSize.`7xl`,
-      lineHeight = BpkLineHeight.`7xl`,
-      letterSpacing = BpkLetterSpacing.Tight,
+  public constructor(defaultFontFamily: FontFamily = FontFamily.SansSerif) : this(
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/tokens/BpkTypography.kt:40]: Can we make the contructor internal?

Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/tokens/BpkTypography.kt:40]: ah yes i believe we can now

**Review by bvitaliyg — APPROVED** [review]

**Pull request merged**